File: src/engine/commands/command.cpp
#include "command.h"

#include <QRegExp>

#include "container.h"
#include "portal.h"
#include "util.h"


#define super QObject

Command::Command(QObject *parent) :
    super(parent) {
}

Command::~Command() {
}

void Command::setDescription(const QString &description) {

    m_description = Util::processHighlights(description);
}

void Command::prepareExecute(Character *player, const QString &command) {

    setCharacter(player);
    setCommand(command);

    m_alias = takeWord();
}

void Command::setCommand(const QString &command) {

    static QRegExp whitespace("\\s+");

    m_words = command.split(whitespace);

    for (int i = 0; i < m_words.length(); i++) {
        QString word = m_words[i];
        if (word.startsWith("\"")) {
            while (i < m_words.length() - 1 && !word.endsWith("\"")) {
                word += " " + m_words[i + 1];
                m_words.removeAt(i);
            }
            if (word.endsWith("\"")) {
                m_words[i] = word.mid(1, word.length() - 2);
            } else {
                m_words[i] = word.mid(1);
            }
        }
    }
}

void Command::setWords(const QStringList &words) {

    m_words = words;
}

void Command::prependWord(const QString &word) {

    m_words.prepend(word);
}

void Command::appendWord(const QString &word) {

    m_words.append(word);
}

bool Command::assertWordsLeft(const QString &noneLeftText) {

    if (!hasWordsLeft()) {
        m_character->send(noneLeftText);
        return false;
    } else {
        return true;
    }
}

QString Command::peekWord() const {

    if (m_words.length() > 0) {
        return m_words.first();
    }

    return QString();
}

QString Command::takeWord(Options options) {

    if (m_words.length() > 0) {
        if (options & IfNotLast && m_words.length() == 1) {
            return QString();
        }
        return m_words.takeFirst();
    }

    return QString();
}

QString Command::takeWord(const char *pattern, Options options) {

    return takeWord(QRegExp(QString(pattern)), options);
}

QString Command::takeWord(const QRegExp &pattern, Options options) {

    if (m_words.length() > 0) {
        if (options & IfNotLast && m_words.length() == 1) {
            return QString();
        }
        if (pattern.exactMatch(m_words[0])) {
            return m_words.takeFirst();
        }
    }

    return QString();
}

GameObjectPtr Command::takeObject(const GameObjectPtrList &pool) {

    ObjectDescription description = takeObjectsDescription();
    if (!description.name.isEmpty()) {
        return objectByDescription(description, pool);
    }

    return GameObjectPtr();
}

GameObjectPtrList Command::takeObjects(const GameObjectPtrList &pool) {

    ObjectDescription description = takeObjectsDescription();
    if (!description.name.isEmpty()) {
        return objectsByDescription(description, pool);
    }

    return GameObjectPtrList();
}

Command::ObjectDescription Command::takeObjectsDescription() {

    takeWord("the", IfNotLast);

    ObjectDescription description;
    description.position = 0;

    if (m_words.isEmpty()) {
        return description;
    }

    description.name = m_words.takeFirst().toLower();

    if (description.name.contains('.')) {
        int index = description.name.indexOf('.');
        description.position = description.name.left(index).toInt();
        if (description.position > 0) {
            description.name = description.name.mid(index + 1);
        }
    } else if (!m_words.isEmpty()) {
        description.position = Util::numericPosition(description.name);
        if (description.position > 0) {
            description.name = m_words.takeFirst().toLower();
        } else {
            description.position = m_words.first().toInt();
            if (description.position > 0) {
                m_words.removeFirst();
            }
        }
    }

    return description;
}

QString Command::takeRest() {

    QString rest = m_words.join(" ");
    m_words.clear();
    return rest;
}

GameObjectPtr Command::objectByDescription(const ObjectDescription &description,
                                           const GameObjectPtrList &pool) {

    GameObjectPtrList objects = objectsByDescription(description, pool);
    if (!objects.isEmpty()) {
        return objects[0];
    }

    return GameObjectPtr();
}

GameObjectPtrList Command::objectsByDescription(const ObjectDescription &description,
                                                const GameObjectPtrList &pool) {

    GameObjectPtrList objects;
    if (description.name == "all") {
        objects = pool;
    } else {
        for (const GameObjectPtr &object : pool) {
            QString name;
            if (object->isPortal()) {
                name = object.unsafeCast<Portal *>()->nameFromRoom(currentRoom());
            } else {
                name = object->name();
            }
            QStringList words = name.toLower().split(' ');
            for (const QString &word : words) {
                if (word.startsWith(description.name)) {
                    objects.append(object);
                    break;
                }
            }
        }
    }
    if (description.position > 0) {
        if (description.position <= (int) objects.length()) {
            GameObjectPtr selected = objects[description.position - 1];
            objects.clear();
            objects.append(selected);
        } else {
            objects.clear();
        }
    }
    return objects;
}

bool Command::requireSome(const GameObjectPtr &object, const QString &tooFewText) {

    if (object.isNull()) {
        m_character->send(tooFewText);
        return false;
    } else {
        return true;
    }
}

bool Command::requireSome(const GameObjectPtrList &objects,
                          const QString &tooFewText) {

    if (objects.length() == 0) {
        m_character->send(tooFewText);
        return false;
    } else {
        return true;
    }
}

bool Command::requireUnique(const GameObjectPtrList &objects,
                            const QString &tooFewText,
                            const QString &tooManyText) {

    if (objects.length() == 0) {
        m_character->send(tooFewText);
        return false;
    } else if (objects.length() > 1) {
        m_character->send(tooManyText);
        return false;
    } else {
        return true;
    }
}

void Command::send(const QString &message, const QString &arg1) {

    send(QString(message).arg(arg1));
}

void Command::send(const QString &message, const QString &arg1, const QString &arg2) {

    send(QString(message).arg(arg1, arg2));
}

void Command::setCharacter(Character *character) {

    m_character = character;
}

File: src/engine/commands/command.js

function Command() {

    this.description = "";
    this.player = null;
    this.currentRoom = null;
    this.words = [];
    this.alias = "";
}

Command.prototype.setDescription = function(description) {

    this.description = Util.processHighlights(description);
};

Command.prototype.execute = null;

Command.prototype.prepareExecute = function(player, command) {

    this.setPlayer(player);
    this.setCommand(command);

    this.alias = this.takeWord();
};

Command.prototype.setPlayer = function(player) {

    this.player = player;
    this.currentRoom = player.currentRoom;
};

Command.prototype.setCommand = function(command) {

    var words = command.trimmed().split(/\s+/);

    for (var i = 0, length = words.length; i < length; i++) {
        var word = words[i];
        if (word.startsWith("\"")) {
            while (i < words.length - 1 && !word.endsWith("\"")) {
                word += " " + words[i + 1];
                words.removeAt(i);
            }
            if (word.endsWith("\"")) {
                words[i] = word.mid(1, word.length - 2);
            } else {
                words[i] = word.mid(1);
            }
        }
    }

    this.words = words;
};

Command.prototype.prependWord = function(word) {

    this.words.splice(0, 0, word);
};

Command.prototype.appendWord = function(word) {

    this.words.push(word);
};

Command.prototype.numWordsLeft = function() {

    return this.words.length;
};

Command.prototype.hasWordsLeft = function() {

    return this.words.length > 0;
};

Command.prototype.assertWordsLeft = function(noneLeftText) {

    if (!this.hasWordsLeft()) {
        this.player.send(noneLeftText);
        return false;
    } else {
        return true;
    }
}

Command.prototype.peekWord = function() {

    return this.words[0];
}

Command.prototype.takeWord = function(pattern, options) {

    if (this.words.length > 0) {
        if (typeof pattern === "string") {
            pattern = new RegExp("^" + pattern + "$");
            if (options & Options.IfNotLast && this.words.length === 1) {
                return "";
            }
            if (pattern.test(this.words[0])) {
                return this.words.takeFirst();
            }
        } else {
            options = pattern;
            if (options & Options.IfNotLast && this.words.length === 1) {
                return "";
            }
            return this.words.takeFirst();
        }
    }

    return "";
};

Command.prototype.takeObject = function(pool) {

    var description = this.takeObjectsDescription();
    if (!description.name.isEmpty()) {
        var objects = this.objectsByDescription(description, pool);
        if (objects.length > 0) {
            return objects[0];
        }
    }

    return null;
};

Command.prototype.takeObjects = function(pool) {

    var description = this.takeObjectsDescription();
    if (!description.name.isEmpty()) {
        return this.objectsByDescription(description, pool);
    }

    return [];
}

Command.prototype.takeObjectsDescription = function(options) {

    this.takeWord("the", Options.IfNotLast);

    var description = { name: "", position: 0, containerName: "", containerPosition: 0 };

    if (this.words.isEmpty()) {
        return description;
    }

    description.name = this.words.takeFirst().toLower();
    description.position = 0;

    if (description.name.contains('.')) {
        var index = description.name.indexOf('.');
        description.position = description.name.left(index).toInt();
        if (description.position > 0) {
            description.name = description.name.mid(index + 1);
        }
    } else if (!this.words.isEmpty()) {
        description.position = Util.numericPosition(description.name);
        if (description.position > 0) {
            description.name = this.words.takeFirst().toLower();
        } else {
            description.position = this.words.first().toInt();
            if (description.position > 0) {
                this.words.removeFirst();
            }
        }
    }

    return description;
};

Command.prototype.takeRest = function() {

    var rest = this.words.join(" ");
    this.words.clear();
    return rest;
};

Command.prototype.objectByDescription = function(description, pool) {

    var objects = this.objectsByDescription(description, pool);
    if (!objects.isEmpty()) {
        return objects[0];
    }

    return null;
}

Command.prototype.objectsByDescription = function(description, pool) {

    var objects = [];
    if (description.name === "all") {
        objects = pool;
    } else {
        for (var i = 0, length = pool.length; i < length; i++) {
            var object = pool[i];
            var name;
            if (object.nameFromRoom) {
                name = object.nameFromRoom(this.currentRoom);
            } else {
                name = object.name;
            }
            var words = name.toLower().split(" ");
            for (var j = 0; j < words.length; j++) {
                var word = words[j];
                if (word.startsWith(description.name)) {
                    objects.push(object);
                    break;
                }
            }
        }
    }
    if (description.position > 0) {
        if (description.position <= objects.length) {
            var selected = objects[description.position - 1];
            objects = [ selected ];
        } else {
            objects.clear();
        }
    }
    return objects;
};

Command.prototype.requireSome = function(objects, tooFewText) {

    if (objects === null || objects.length === 0) {
        this.send(tooFewText);
        return false;
    } else {
        return true;
    }
};

Command.prototype.requireUnique = function(objects, tooFewText, tooManyText) {

    if (objects.length === 0) {
        this.send(tooFewText);
        return false;
    } else if (objects.length > 1) {
        this.send(tooManyText);
        return false;
    } else {
        return true;
    }
};

Command.prototype.send = function(message, arg1, arg2) {

    if (message.contains("%1")) {
        message = message.replace("%1", arg1);
        if (message.contains("%2")) {
            message = message.replace("%2", arg2);
        }
    }

    this.player.send(message);
};
